### PR TITLE
feat: generate events on pods when apply Kong configurations

### LINKED ISSUE #4003: Generate a k8s Event when Kong dbless POST /config gets a 500 response
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

https://github.com/Kong/kubernetes-ingress-controller/issues/3928 uncovered that if KIC gets a 500 response from `POST /config`, no `Event` gets created by KIC. 
Also, when KIC call `POST /config` of admin API and got a 400 result, but KIC could not parse it into resource errors, no `Event` gets created,

### Proposed Solution

KIC creates an `Event` with `Warning` type in case it received a 500 response or unparsable 400 response trying to `POST /config` to Kong.
When KIC gets 200 response, creates a `Normal` event to tell that the application f kong configuration succedded.

This event can be associated with:
- the KIC pod (using downward API to obtain the pod namespaced name)
- the associated `IngressClass` and/or `Gateway` (note that the 500 is not specific to whether Gateway API or Ingress API is in use

### Additional information

_No response_

### Acceptance Criteria

- [ ] If KIC gets a 500 response or an unparsable 400 to a `POST /config` request, KIC creates a `Warning` Kubernetes Event informing about that.
- [ ] If KIC gets a 200 resonse, create a `Normal` event for it.

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

Generates an event attached to KIC pod on KIC applying configuration to Kong.
When KIC successfully applied configurations, attach a `Normal` event;
When applying fails, generate a `Warming` event with error message. 

**Which issue this PR fixes**:

fixes #4003

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/kong_client.go [hunk 1]
```diff
@@ -484,6 +485,7 @@ func (c *KongClient) sendToClient(
 	)
 
 	c.recordResourceFailureEvents(entityErrors, KongConfigurationApplyFailedEventReason)
+	c.recordApplyConfigurationEvents(err)
```
Comment (pmalek) [inline internal/dataplane/kong_client.go:488]: Can we include the client address in the event as well? The information is readily available in scope of this function.

File: internal/dataplane/sendconfig/inmemory.go [hunk 2]
```diff
@@ -57,7 +57,8 @@ func (s UpdateStrategyInMemory) Update(ctx context.Context, targetState ContentW
 	}
 
 	flattenErrors := shouldUseFlattenedErrors()
-	if errBody, err := s.configService.ReloadDeclarativeRawConfig(ctx, bytes.NewReader(config), true, flattenErrors); err != nil {
+	errBody, err := s.configService.ReloadDeclarativeRawConfig(ctx, bytes.NewReader(config), true, flattenErrors)
```
Comment (programmer04) [inline internal/dataplane/sendconfig/inmemory.go:60]: When `err` is not `nil` the other returned value in this case `errBody` is treated as valid and not something to discard, it's not idiomatic Go to do like that

File: internal/dataplane/kong_client.go [hunk 3]
```diff
@@ -617,3 +619,29 @@ func (c *KongClient) recordResourceFailureEvents(resourceFailures []failures.Res
...
+
+	if err != nil {
+		eventType = corev1.EventTypeWarning
+		reason = "failed to apply configuration to Kong"
+		message = err.Error()
+	}
+
+	podNN, getPodErr := util.GetPodNN()
+	if getPodErr != nil {
+		c.logger.Error(getPodErr)
```
Comment (czeslavo) [inline internal/dataplane/kong_client.go:636]:
```suggestion
		c.logger.WithError(getPodErr).Error("failed to resolve controller's pod to attach the apply configuration event to")
```

File: internal/dataplane/sendconfig/inmemory.go (hunk 2 above)
Comment (czeslavo) [reply internal/dataplane/sendconfig/inmemory.go:60]: Yeah, I agree that `ReloadDeclarativeRawConfig` could have a better interface and return just a single error, allowing inspection of the error body by casting the error to some predefined custom error struct or something similar.

But in this case, I think we'll have to ignore it as `ReloadDeclarativeRawConfig` is not defined in KIC but `go-kong` and it cannot be improved in this PR.

File: internal/dataplane/kong_client.go (hunk 1 above)
Comment (randmonkey) [reply internal/dataplane/kong_client.go]: done.

File: internal/dataplane/kong_client.go (hunk 3 above)
Comment (randmonkey) [reply internal/dataplane/kong_client.go]: done

**Comment by randmonkey** [comment]
> Can we add a changelog entry pray ?

Added CHANGELOG item.

File: internal/dataplane/kong_client.go [hunk 4]
```diff
@@ -484,6 +485,10 @@ func (c *KongClient) sendToClient(
 	)
 
 	c.recordResourceFailureEvents(entityErrors, KongConfigurationApplyFailedEventReason)
+	// REVIEW: do we record the event for uploading config to Konnect?
```
Comment (pmalek) [inline internal/dataplane/kong_client.go:488]: I'm not sure about this one. @czeslavo WDYT?

File: internal/dataplane/kong_client.go [hunk 5]
```diff
@@ -617,3 +622,30 @@ func (c *KongClient) recordResourceFailureEvents(resourceFailures []failures.Res
...
+
+// recordApplyConfigurationEvents records event attached to KIC pod after KIC applied Kong configuration.
+func (c *KongClient) recordApplyConfigurationEvents(err error, rootURL string) {
+	eventType := corev1.EventTypeNormal
+	reason := "KongConfigApplySucceeded"
+	message := fmt.Sprintf("successfully applied Kong configuration to %s", rootURL)
+
+	if err != nil {
+		eventType = corev1.EventTypeWarning
+		reason = "KongConfigApplyFailed"
```
Comment (czeslavo) [inline internal/dataplane/kong_client.go:634]: Isn't it duplicating the already existing `KongConfigurationApplyFailed` event reason?

File: internal/dataplane/kong_client.go (hunk 4 above)
Comment (czeslavo) [reply internal/dataplane/kong_client.go:488]: I'm not sure. One reason to not record such events for Konnect I can see is that if a user would like setup some monitoring based on the these events directly, it won't be programatically easy (without some heuristics like verifying the client URL) to tell if it is an event for a regular gateway or Konnect. 

I'd say that if we'd like to emit something for Konnect, we should rather make that a separate event reason, e.g. `KonnectConfigurationSyncFailed`, but I'm not sure if it's in the scope of this issue.

File: internal/dataplane/kong_client.go [hunk 6]
```diff
@@ -617,3 +622,30 @@ func (c *KongClient) recordResourceFailureEvents(resourceFailures []failures.Res
 		}
 	}
 }
+
+// recordApplyConfigurationEvents records event attached to KIC pod after KIC applied Kong configuration.
+func (c *KongClient) recordApplyConfigurationEvents(err error, rootURL string) {
+	eventType := corev1.EventTypeNormal
+	reason := "KongConfigApplySucceeded"
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/kong_client.go]: done.

File: internal/dataplane/kong_client.go (hunk 5 above)
Comment (randmonkey) [reply internal/dataplane/kong_client.go]: Done

File: internal/dataplane/kong_client.go (hunk 4 above)
Comment (randmonkey) [reply internal/dataplane/kong_client.go]: Decided not to include updating Konnect. @czeslavo  please create an issue if we need to add events with different reasons on updating Konnect.

File: CHANGELOG.md [hunk 7]
```diff
@@ -123,6 +123,10 @@ Adding a new version? You'll need three changes:
   now accounts for potential failures in synchronizing configuration with Konnect's
   Runtime Group Admin API.
   [#4029](https://github.com/Kong/kubernetes-ingress-controller/pull/4029)
+- Record an event attached to KIC pod after applying configuration to Kong. If
+  the applying succeeded, a `Normal` event is recorded. If the applying failed,
+  a `Warning` event is recorded.
+  [#4054](https://github.com/Kong/kubernetes-ingress-controller/pull/4054)
```
Comment (czeslavo) [inline CHANGELOG.md:129]:
```suggestion
- Record an event attached to KIC pod after applying configuration to Kong. If
  the applying succeeded, a `Normal` event with `KongConfigurationSucceeded` 
  reason is recorded. If the applying failed, a `Warning` event with 
  `KongConfigurationApplyFailed` reason is recorded.
  [#4054](https://github.com/Kong/kubernetes-ingress-controller/pull/4054)
```

I think that might be useful to tell users what reasons they should expect so they can filter events using those.

**Review by czeslavo — APPROVED** [review]

**Pull request merged**